mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+    }
+
+    // Retrieve specified config object from "MetaData" partition.
+    pub async fn get_config_object(
+        &self,
+        metadata: &NexusMetaData,
+        n: u32,
+    ) -> Result<Option<NexusConfig>, MetaDataError> {
+        if n < metadata.header.used_entries {
+            return Ok(Some(self.probe_config_object(metadata, n).await?));
+        }
+        Ok(None)
+    }
+
+    // Retrieve latest config object from "MetaData" partition.
+    pub async fn get_latest_config_object(
+        &self,
+        metadata: &NexusMetaData,
+    ) -> Result<Option<NexusConfig>, MetaDataError> {
+        if metadata.header.used_entries > 0 {
+            let n = metadata.header.used_entries - 1;
+            return Ok(Some(self.probe_config_object(metadata, n).await?));
+        }
+        Ok(None)
+    }
+
+    // Remove specified config object from "MetaData" partition.
+    pub async fn delete_config_object(
+        &mut self,
+        metadata: &mut NexusMetaData,
+        n: u32,
+    ) -> Result<(), MetaDataError> {
+        if metadata.index.len() != metadata.header.used_entries as usize {
+            return Err(MetaDataError::IndexInconsistent {});
+        }
+
+        if n < metadata.header.used_entries {
+            metadata.index.remove(n as usize);
+            metadata.header.used_entries -= 1;
+            return self.sync_metadata(metadata).await;
+        }
+        Ok(())
+    }
+
+    // Append a new config object to "MetaData" partition.
+    pub async fn append_config_object(
+        &mut self,
+        metadata: &mut NexusMetaData,
+        config: &NexusConfig,
+        now: &SystemTime,
+    ) -> Result<(), MetaDataError> {
+        if metadata.index.len() != metadata.header.used_entries as usize {
+            return Err(MetaDataError::IndexInconsistent {});
+        }
+
+        if NexusChild::fragmented(metadata.header.data_start, &metadata.index)?
+        {
+            if let Err(error) = self.compact(metadata).await {
+                warn!("Error compacting MetaData: {}", error);
+            }
+        }
+
+        self.write_config_object(metadata, config, now).await?;
+        self.sync_metadata(metadata).await
+    }
+}
+
+impl NexusConfig {
+    pub fn from_slice(buf: &[u8]) -> Result<NexusConfig, Error> {
+        let mut reader = Cursor::new(buf);
+        let config: NexusConfig = deserialize_from(&mut reader)?;
+        Ok(config)
+    }
+}
+
+// ChildError already has a variant with a MetaDataError source.
+// Attempting to add a variant to MetaDataError with a ChildError source
+// creates a circular definition, resulting in the compiler error:
+//   "cycle detected when processing ChildError"
+// This is resolved by providing an explicit conversion.
+impl std::convert::From<ChildError> for MetaDataError {
+    fn from(error: ChildError) -> MetaDataError {
+        MetaDataError::NexusChildError {
+            error: format!("{}", error),
+        }
+    }
+}
```

Comment:
Cycles like this usually indicate that there is something wrong with the architecture. It would be good to have a look again if the code is split optimally between the modules. For example there are NexusChild methods implemented in metadata module using metadata error type. I think that's the first confusing thing. I would expect NexusChild methods being defined in nexus_child module along with the nexus-child struct and using nexus-child-error instead of metadata-error. But I don't have a good understanding of the whole feature. Perhaps we can take the discussion offline to clarify things.